src/main/java/org/opensearch/security/privileges/ActionPrivileges.java

```diff
@@ -0,0 +1,1146 @@
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.privileges;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.cluster.metadata.DataStream;
+import org.opensearch.cluster.metadata.IndexAbstraction;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.common.unit.ByteSizeUnit;
+import org.opensearch.core.common.unit.ByteSizeValue;
+import org.opensearch.security.resolver.IndexResolverReplacer;
+import org.opensearch.security.securityconf.FlattenedActionGroups;
+import org.opensearch.security.securityconf.impl.SecurityDynamicConfiguration;
+import org.opensearch.security.securityconf.impl.v7.RoleV7;
+import org.opensearch.security.support.WildcardMatcher;
+
+import com.selectivem.collections.CheckTable;
+import com.selectivem.collections.CompactMapGroupBuilder;
+import com.selectivem.collections.DeduplicatingCompactSubSetBuilder;
+import com.selectivem.collections.ImmutableCompactSubSet;
+
+/**
+ * This class converts role configuration into pre-computed, optimized data structures for checking privileges.
+ * <p>
+ * With the exception of the statefulIndex property, instances of this class are immutable. The life-cycle of an
+ * instance of this class corresponds to the life-cycle of the role and action group configuration. If the role or
+ * action group configuration is changed, a new instance needs to be built.
+ */
+public class ActionPrivileges extends ClusterStateMetadataDependentPrivileges {
+
+    /**
+     * This setting controls the allowed heap size of the precomputed index privileges (in the inner class StatefulIndexPrivileges).
+     * If the size of the indices exceed the amount of bytes configured here, it will be truncated. Privileges evaluation will
+     * continue to work correctly, but it will be slower.
+     * <p>
+     * This settings defaults to 10 MB. This is a generous limit. Experiments have shown that an example setup with
+     * 10,000 indices and 1,000 roles requires about 1 MB of heap. 100,000 indices and 100 roles require about 9 MB of heap.
+     * (Of course, these numbers can vary widely based on the actual role configuration).
+     * <p>
+     * The setting plugins.security.privileges_evaluation.precomputed_privileges.include_indices can be used to control
```

Comment:
The setting cannot be updated dynamically. 

To be honest, I was wondering about removing this again:

As it turns out, the resulting instances are really small - even for 10k indices and 1000 roles, we are below 1 MB of heap size - which is virtually nothing.

What do you think, does bring an actual benefit - or does it just cause unnecessary friction (code maintenance, additional docs which need to be read and understood by users, etc).

Reply:
If we think that it is not needed then we can not add it TBH. We can add it later, but we cannot add it now and remove it later. I'm not have of case where clusters are ginormous and require specifying indices to be pre-computed.

Reply:
Good point!

Reply:
Removed in https://github.com/opensearch-project/security/pull/4380/commits/490cef90902d4130d14780098f2dd10f500055a1 ... I kept "plugins.security.privileges_evaluation.precomputed_privileges.max_heap_size" because I like to have a strong abort condition for that loop, though.